ASAA-17 Arch Demo Commit List

LuisRocketRangel opened the pull request:
Requirements:
-Update Nav Drawer to conditionally show source and commit options
-Simple List of Commits from Repo
-Document Process to add Tabs to NavDrawer in the README.md
-Create defaults in Routes.kt for RepoBrowser and RepoFile to help with matching

Solution:
-Updated code to get commits from the Bitbucket API
-Added new composables for the Commit List Screen and Commit List Item
-Added a new route for commit screen
-Implemented default values for routes with args to help with matching
-Implemented nestedMenuItems in NavItemState to support nesting menu lists
-Added assets relating to commits and source items
-Updated the README.md to include documentation on adding items to the Nav Drawer
![Screenshot_20220804-144236](https://user-images.githubusercontent.com/83731304/182942109-031b4c04-2185-4855-872e-50c3917d88a2.png)
![Screenshot_20220804-144242](https://user-images.githubusercontent.com/83731304/182942113-53762e66-dde7-4dd3-b577-d94336bcfb0c.png)
![Screenshot_20220804-144251](https://user-images.githubusercontent.com/83731304/182942117-4355a867-e3f2-4299-98cf-d696446c03bb.png)
![Screenshot_20220804-145741](https://user-images.githubusercontent.com/83731304/182942121-bb1ef918-afc9-472e-98e9-464b7bc2825e.png)

LuisRocketRangel committed:
ASAA-17 Arch Demo Commit List

Requirements:
-Update Nav Drawer to conditionally show source and commit options
-Simple List of Commits from Repo
-Document Process to add Tabs to NavDrawer in the README.md
-Create defaults in Routes.kt for RepoBrowser and RepoFile to help with matching

Solution:
-Updated code to get commits from the Bitbucket API
-Added new composables for the Commit List Screen and Commit List Item
-Added a new route for commit screen
-Implemented default values for routes with args to help with matching
-Implemented nestedMenuItems in NavItemState to support nesting menu lists
-Added assets relating to commits and source items
-Updated the README.md to include documentation on adding items to the Nav Drawer

BottleRocket-Colin commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/MainNavGraph.kt:
```diff
@@ -182,6 +184,23 @@ private fun ComposeActivity.repositoryBrowserComposable(navGraphBuilder: NavGrap
+//         controls.topLevel = true
+//     }
+// }
+
```
If we don't need this commented code anymore, go ahead and delete.

BottleRocket-Colin commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/Routes.kt:
```diff
@@ -12,13 +12,14 @@ object Routes {
+    const val Commits = "commits"
 
-    fun RepositoryBrowser(data: RepositoryBrowserData) =
+    fun RepositoryBrowser(data: RepositoryBrowserData = RepositoryBrowserData("{repoName}", "{folderHash}", "{folderPath}")) =
```
Can we use these new default values in MainNavGraph for the route definition?  If so, that would be great addition.

LuisRocketRangel replied:
I can look into that, it works when doing the current route check in Compose Activity

BottleRocket-Colin commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryCommitViewModel.kt:
```diff
@@ -0,0 +1,71 @@
+    // State
+    private val srcCommits = MutableStateFlow<List<CommitDto>>(emptyList())
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    var currentRepoName: String = ""
```
Could we make this a mutable state flow?  Then:

init {
  launchIO {
    currentRepoName.collect { repoName -> getRepoCommits(repoName)
}
}

BottleRocket-Colin replied:
This will allow us to make getRepoCommits private and then anytime currentRepoName changes value it will fetch the correct info.

LuisRocketRangel replied:
Got it! I sloppily wrote this down in my notes. I'll get this done!

BottleRocket-Colin commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryCommitsComposable.kt:
```diff
@@ -0,0 +1,25 @@
+
+fun ComposeActivity.repositoryCommitsComposable(navGraphBuilder: NavGraphBuilder) {
+    navGraphBuilder.composable(
+        "commits"
```
Use Routes.  object for this hardcoded value, if possible.

BottleRocket-Colin commented on compose/src/main/java/com/bottlerocketstudios/compose/navdrawer/NavDrawer.kt:
```diff
@@ -71,6 +72,13 @@ fun ColumnScope.NavDrawer(state: NavDrawerState) {
         NavDrawerItem(state = item)
+        Column(modifier = Modifier.padding(start = Dimens.grid_4)) {
+            ProvideTextStyle(value = MaterialTheme.typography.body1) {
+                for (nestedItem in item.nestedMenuItems) {
```
Nice use of style provider!

BottleRocket-Colin commented on data/src/main/java/com/bottlerocketstudios/brarchitecture/data/model/RepoFile.kt:
```diff
@@ -13,5 +13,5 @@ data class RepoFile(
     @Json(name = "attributes") val attributes: List<String>?,
     @Json(name = "size") val size: Int?,
-    @Json(name = "commit") val commit: Commit?
+    @Json(name = "commit") val commit: CommitDto?
```
Thanks for making this change. :)

BottleRocket-Colin reviewed (changes requested):
Some minor updates to be made.  Excellent first foray into compose.

jaredstockton commented:
> Added new composables for the Commit List Screen and Commit List Item

@LuisRocketRangel Would you mind adding a few sample screenshots of these screens with test data?

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/ComposeActivity.kt:
```diff
@@ -124,12 +126,30 @@ class ComposeActivity : ComponentActivity() {
-                selected = currentRoute == Routes.Home
+                selected = getTopRoute(currentRoute) == Routes.Home,
+                nestedMenuItems = if (showHomeSubList && getTopRoute(currentRoute) == Routes.Home) listOf(
+                    NavItemState(
+                        icon = R.drawable.ic_source,
+                        itemText = R.string.home_nav_source,
+                        selected = getTopRoute(currentRoute) == Routes.Home,
+                    ) {
+                        scaffoldState.drawerState.close()
+                        navController.navigate(Routes.RepositoryBrowser(RepositoryBrowserData(repoName = activityViewModel.selectedRepo.value.name ?: "")))
+                    },
+                    NavItemState(
+                        icon = R.drawable.ic_commit,
+                        itemText = R.string.home_nav_commits,
+                        selected = getTopRoute(currentRoute) == Routes.Home,
+                    ) {
+                        scaffoldState.drawerState.close()
+                        navController.navigate(Routes.Commits)
+                    }
```
@LuisRocketRangel It seems like the selected state should be either source or commit as it is unclear which one the user currently has selected when the drawer is open when they're both selected (and using red text)
![Screenshot_20220804_092237](https://user-images.githubusercontent.com/6143792/182873151-8a9da1bd-19d4-4429-b6c4-7047986f31ec.png)
.

Zeplin is unclear on this behavior as I don't see an example of a submenuitem being selected. In that case, I'd recommend updating this logic to only show selected state for the actual composable screen rendered (either source or commit)

BottleRocket-Colin replied:
@br-jared-stockton / @LuisRocketRangel  - I agree with Jared.  This is a place where we can diverge from Zeplin and implement a more logical approach.  This should also allow you to remove the find top level nav item function and clean things up a bit. :)

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/ComposeActivity.kt:
```diff
@@ -151,4 +171,13 @@ class ComposeActivity : ComponentActivity() {
+        when (route) {
+            Routes.Home,
+            Routes.Commits,
+            Routes.RepositoryBrowser()
```
@LuisRocketRangel Calling `Routes.RepositoryBrowser()` uses a default value as no argument are passed in. If this function is still necessary after refactoring based on the selection state comment, then this logic should also be updated as a non-default RepositoryBrowser route won't match. 

@BottleRocket-Colin This ties into the broader concept of determining the current screen. Would be nice to have a reference of how we can do that especially when we use templated style routes (where we effectively plug in path and query values into the route)

BottleRocket-Colin replied:
@br-jared-stockton Agreed.  I'm working on a more generalized version of nav drawer in PathFinder as we type. :).  Once I get that prototyped.  I'll move it to LaunchPad and then retrofit back into Arch Demo.

BottleRocket-Colin replied:
@br-jared-stockton  - Just as an FYI, I believe that the "current route" passed in to navigation composable will always be the generalized version with brackets and arguments will already be parsed out.  At least this is behavior that @LuisRocketRangel observed.  A moot point since this code is going away, but worth noting for the future.

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/MainNavGraph.kt:
```diff
@@ -182,6 +184,23 @@ private fun ComposeActivity.repositoryBrowserComposable(navGraphBuilder: NavGrap
+//         controls.title = repoName
+//         controls.topLevel = true
+//     }
+// }
```
@LuisRocketRangel Delete this commented block OR uncomment and use it :)

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryCommitStateConverter.kt:
```diff
@@ -0,0 +1,15 @@
+    commitItems = uiModels.collectAsState()
+    // onRepositoryItemClicked = { item ->
+    //     onRepoItemClicked(item)
+    // }
```
@LuisRocketRangel Delete the commented block OR uncomment it if just done for testing

jaredstockton commented on app/src/main/res/values/strings.xml:
```diff
@@ -13,6 +13,10 @@
 
+    <!-- Repository Commits Screen -->
+    <string name="error_loading_commits" tools:ignore="ExtraTranslation">Error loading commits</string>
+    <string name="commit_message" tools:ignore="ExtraTranslation">Message: $s</string>
```
@LuisRocketRangel Append `_format` to the string name to indicate that this string is a format string and requires extra care at the use site (passing in format values)

jaredstockton commented on compose/src/main/java/com/bottlerocketstudios/compose/repository/RepositoryCommitScreenState.kt:
```diff
@@ -0,0 +1,10 @@
+    val path: State<String>,
+    val itemCount: State<Int>,
+    val commitItems: State<List<RepositoryCommitItemUiModel>>,
+    // val onRepositoryItemClicked: (RepositoryCommitItemUiModel) -> Unit
```
@LuisRocketRangel Delete commented code if not used

jaredstockton commented on README.md:
```diff
@@ -46,6 +46,22 @@ The project is using the Gradle Versions plugin to do a lot of heavy lifting her
 
 *Note that you'll still need to manually check for dependency updates on all buildscript plugins (see `Dependencies.kt` -> `Config.BuildScriptPlugins` properties), including the Gradle Versions plugin.*
 
+##Adding Tabs to the Nav Drawer
```
@LuisRocketRangel nit: Add a space between the heading indicators and the heading text
```suggestion
## Adding Tabs to the Nav Drawer
```

jaredstockton commented on README.md:
```diff
@@ -46,6 +46,22 @@ The project is using the Gradle Versions plugin to do a lot of heavy lifting her
 *Note that you'll still need to manually check for dependency updates on all buildscript plugins (see `Dependencies.kt` -> `Config.BuildScriptPlugins` properties), including the Gradle Versions plugin.*
 
+##Adding Tabs to the Nav Drawer
+This project uses the `generateNavDrawerItems` method in the `ComposeActivity` to draw items in the Nav Drawer.
```
Thanks for this helper text!

jaredstockton reviewed (changes requested):
Nice job overall, just a few comments to address!

LuisRocketRangel commented:
> > Added new composables for the Commit List Screen and Commit List Item
> 
> @LuisRocketRangel Would you mind adding a few sample screenshots of these screens with test data?

Hey @br-jared-stockton Where do I place the screenshots?

LuisRocketRangel committed:
ASAA-17 PR Feedback

Requirements:
-Implement MutableStateFlow for currentRepoName in RepCommitVM
-Correct Routes
-Change selected behavior in the Nav Drawer to highlight the selected submenu only
-Removed commented out code

Solution:
-Implemented the MutableStateFlow to collect changes in currentRepoName and to make getCommits private in the VM init block
-Fixed routes
-Changed the selected nav item behavior to highlight sub menu items
-Removed unused code
-Fixed small nits

jaredstockton commented:
> > > Added new composables for the Commit List Screen and Commit List Item
> > 
> > 
> > @LuisRocketRangel Would you mind adding a few sample screenshots of these screens with test data?
> 
> Hey @br-jared-stockton Where do I place the screenshots?

Editing the description and dragging/dropping into the text input would be ideal. If that doesn't work for some reason, you could just create a new comment and drag/drop them there :)

LuisRocketRangel committed:
ASAA-17 PR Feedback

Requirements:
-Clean up Leftover code

Solution:
-Cleaned UP Leftover code

BottleRocket-Colin commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryCommitViewModel.kt:
```diff
@@ -0,0 +1,75 @@
+
+    // State
+    private val srcCommits = MutableStateFlow<List<CommitDto>>(emptyList())
+    var currentRepoName = MutableStateFlow("")
```
This should be a val

BottleRocket-Colin commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryCommitViewModel.kt:
```diff
@@ -0,0 +1,75 @@
+    init {
+        launchIO {
+            currentRepoName.collect {
+                repoName ->
```
this could/should read : 
currentRepoName.collect { repoName ->

on one line.

BottleRocket-Colin reviewed (changes requested):
Down to 2 very minor nit picks.  Great stuff so far!  Keep it up.

LuisRocketRangel committed:
ASAA-17 PR Feedback II

Requirements:
-Minor Nits

Solution:
-Fix the Nits

jaredstockton reviewed (approved):
Thanks for the help on this one @LuisRocketRangel!
